onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc

```diff
@@ -0,0 +1,356 @@
+        const auto& input_name = model->GetInputs()[i];
+        const auto& model_input_type = model->GetInputType(input_name);
+
+        auto input_idx = model->GetMappedInputIdx(input_name);
+        const OrtValue* input_tensor = ort.KernelContext_GetInput(context, input_idx);
+        const auto tensor_info = ort.GetTensorTypeAndShape(input_tensor);
+        const auto& tensor_shape = ort.GetTensorShape(tensor_info);
+        std::vector<uint32_t> dimensions;
+        for (const auto& dim : tensor_shape)
+          dimensions.push_back(static_cast<uint32_t>(dim));
+
+        // it is possible that the input has the detailed size while
+        // the model has an operand with unknown size, use the size
+        // of the actual input
+        OperandType type(model_input_type.type, dimensions,
+                         model_input_type.operandType.scale,
+                         model_input_type.operandType.zeroPoint);
+
+        ORT_ENFORCE(type.dimensions == model_input_type.dimensions ||
+                        model_input_type.GetOperandBlobByteSize() == 0,
+                    "The actual input dimanesions should match the model input "
+                    "dimensions, or model input dimension has 0 (dynamic)");
+
+        void* inputBuffer = const_cast<void*>(ort.GetTensorData<void>(input_tensor));
+        inputs.push_back({inputBuffer, std::move(type)});
+
+        ort.ReleaseTensorTypeAndShapeInfo(tensor_info);
+      }
+
+      model->SetInputBuffers(inputs);
+      std::vector<nnapi::Model::InputOutputBuffer> outputs;
+      outputs.reserve(num_outputs);
+      for (size_t i = 0; i < num_outputs; i++) {
+        const auto output_name = model->GetOutputs()[i];
+        const auto model_output_type = model->GetOutputType(output_name);
+        const auto output_shape = model_output_type.dimensions;
+
+        std::vector<int64_t> int64_output_shape(output_shape.begin(),
+                                                output_shape.end());
+        auto output_idx = model->GetMappedOutputIdx(output_name);
+        auto* output_tensor = ort.KernelContext_GetOutput(context, output_idx,
+                                                          int64_output_shape.data(),
+                                                          int64_output_shape.size());
+
+        void* output_buffer = nullptr;
+        switch (model_output_type.type) {
+          case Type::TENSOR_FLOAT32:
+            output_buffer = ort.GetTensorMutableData<float>(output_tensor);
+            break;
+          case Type::TENSOR_INT32:
+            output_buffer = ort.GetTensorMutableData<int32_t>(output_tensor);
+            break;
+          default:
+            ORT_THROW("Unsupported output type: " + TypeToStr(model_output_type.type));
+            break;
+        }
+
+        ORT_ENFORCE(model_output_type.GetOperandBlobByteSize() != 0,
+                    "We do not support dynamic output shape for now");
+
```

Comment:
ORT_ENFORCE will throw. If it's something we don't support should we exit more gracefully?

Reply:
Ideally we should mark this model is not supported at all in GetCapability() such that the model will not run on this EP, but it is not straightforward to tell whether the output shape is dynamic unless you run construct the model once.

The dynamic output shape is a TODO which will be addressed later

Reply:
If you throw if won't fallback to the cpu EP - execution will exit.